enzyme/Enzyme/Enzyme.cpp

```diff
@@ -377,102 +372,99 @@ static Value *adaptReturnedVector(CallInst *CI, Value *diffret,
+        Value *sgep = Builder.CreateStructGEP(retType, ret, i);
 #else
-          Value *sgep = Builder.CreateStructGEP(sret, i);
+        Value *sgep = Builder.CreateStructGEP(ret, i);
 #endif
-          Builder.CreateStore(Builder.CreateExtractValue(diffret, {i}), sgep);
-        }
-      } else {
-        auto &DL = fn->getParent()->getDataLayout();
-        if (DL.getTypeSizeInBits(sret_ty) !=
-            DL.getTypeSizeInBits(diffret->getType())) {
-          EmitFailure("IllegalReturnCast", CI->getDebugLoc(), CI,
-                      "Cannot cast return type of gradient ",
-                      *diffret->getType(), *diffret, ", to desired type ",
-                      *stype->getPointerElementType());
-          return false;
-        }
-        Builder.CreateStore(
-            diffret, Builder.CreatePointerCast(
-                         sret, PointerType::get(diffret->getType(),
-                                                stype->getAddressSpace())));
+        Builder.CreateStore(Builder.CreateExtractValue(diffret, {i}), sgep);
       }
-    } else if (mode == DerivativeMode::ReverseModePrimal) {
-      auto &DL = fn->getParent()->getDataLayout();
-      if (DL.getTypeSizeInBits(CI->getType()) >=
-          DL.getTypeSizeInBits(diffret->getType())) {
-        IRBuilder<> EB(&CI->getParent()->getParent()->getEntryBlock().front());
-        auto AL = EB.CreateAlloca(CI->getType());
-        Builder.CreateStore(
-            diffret, Builder.CreatePointerCast(
-                         AL, PointerType::getUnqual(diffret->getType())));
+      CI->eraseFromParent();
+      return true;
+    }
+
+    if (DL.getTypeSizeInBits(retType) >= DL.getTypeSizeInBits(diffretType)) {
+      Builder.CreateStore(
+          diffret,
+          Builder.CreatePointerCast(ret, PointerType::getUnqual(diffretType)));
+      CI->eraseFromParent();
+      return true;
+    }
+  }
+
+  if (mode == DerivativeMode::ReverseModePrimal &&
+      DL.getTypeSizeInBits(retType) >= DL.getTypeSizeInBits(diffretType)) {
+    IRBuilder<> EB(CI->getFunction()->getEntryBlock().getFirstNonPHI());
+    auto AL = EB.CreateAlloca(retType);
+    Builder.CreateStore(diffret, Builder.CreatePointerCast(
+                                     AL, PointerType::getUnqual(diffretType)));
 #if LLVM_VERSION_MAJOR > 7
-        Value *cload = Builder.CreateLoad(CI->getType(), AL);
+    Value *cload = Builder.CreateLoad(retType, AL);
 #else
-        Value *cload = Builder.CreateLoad(AL);
+    Value *cload = Builder.CreateLoad(AL);
 #endif
-        CI->replaceAllUsesWith(cload);
-      } else {
-        EmitFailure("IllegalReturnCast", CI->getDebugLoc(), CI,
-                    "Cannot cast return type of gradient ", *diffret->getType(),
-                    *diffret, ", to desired type ", *CI->getType());
-        return false;
-      }
-    } else {
+    CI->replaceAllUsesWith(cload);
+    CI->eraseFromParent();
+    return true;
+  }
 
```

Comment:
Why does this need the mode != primal?

Reply:
This if has been there for a [while](https://github.com/EnzymeAD/Enzyme/blob/25bf34184088629743bc85bb7d48bec716a36a2a/enzyme/Enzyme/Enzyme.cpp#L577). I tried to keep this change a non-functional change.